arangod/Replication2/StateMachines/Document/DocumentStateTransactionHandler.cpp

```diff
@@ -101,79 +108,101 @@ auto DocumentStateTransactionHandler::getTrx(TransactionId tid)
   return it->second;
 }
 
-auto DocumentStateTransactionHandler::applyEntry(DocumentLogEntry doc)
-    -> Result {
-  if (doc.operation == OperationType::kAbortAllOngoingTrx) {
-    _transactions.clear();
-    return Result{};
-  }
+void DocumentStateTransactionHandler::setTrx(
+    TransactionId tid, std::shared_ptr<IDocumentStateTransaction> trx) {
+  auto [_, isInserted] = _transactions.emplace(tid, std::move(trx));
+  TRI_ASSERT(isInserted) << "Transaction " << tid << " already exists";
+}
 
-  TRI_ASSERT(doc.tid.isFollowerTransactionId());
-
-  try {
-    auto trx = ensureTransaction(doc);
-    TRI_ASSERT(trx != nullptr);
-    switch (doc.operation) {
-      case OperationType::kInsert:
-      case OperationType::kUpdate:
-      case OperationType::kReplace:
-      case OperationType::kRemove:
-      case OperationType::kTruncate: {
-        auto opRes = trx->apply(doc);
-        auto res = opRes.fail() ? opRes.result
-                                : makeResultFromOperationResult(opRes, doc.tid);
-        if (res.fail() && shouldIgnoreError(opRes)) {
-          LoggerContext logContext =
-              LoggerContext(Logger::REPLICATED_STATE)
-                  .with<logContextKeyDatabaseName>(_gid.database)
-                  .with<logContextKeyLogId>(_gid.id);
-
-          LOG_CTX("0da00", INFO, logContext)
-              << "Result ignored while applying transaction " << doc.tid
-              << " with operation " << int(doc.operation) << " on shard "
-              << doc.shardId << ": " << res;
+auto DocumentStateTransactionHandler::applyEntry(ReplicatedOperation operation)
+    -> Result try {
+  return std::visit(
+      [this, &operation](auto&& op) -> Result {
```

Comment:
Same as before, we probably want to consider using override visitor here.
With this many cases it gets a bit hard to visually see that all `else if constexpr` are correct, and we not only have a single if at one place.

Reply:
My problem is that I'm trying to have these `static_assert(always_false_v<T>, "Unhandled operation type. This should never happen.");` help me make sure that I handle all cases at compile time. In the end, the overload solution would work fine for me, I don't have a strong preference on how to achieve this visitor thing. We could bring that as a topic of debate to one of our team meetings and decide upon a common approach.